### ISSUE TITLE
Losing connection to DB

### ISSUE THREAD

**Issue by edgarcosta** [LMFDB/lmfdb#1209]
Hello,

At the moment the client quits whenever it loses the connection to the DB.
In my opinion it should not.

Best,
Edgar

**Comment by JohnCremona** [comment]
I take you do not just mean that this is an inconvenience for developers, but actually a problem for the live website?

**Comment by jvoight** [comment]
Give us a minimal nonworking example or what you'd like to have changed and what the implications would be.  Also, an indication of how serious this is.

**Comment by edgarcosta** [comment]
If the client loses the connection to the mongodb server you get an "Internal Server Error" from gunicorn.

Minimal example, http://lmfdb.org/acknowledgment

**Comment by jvoight** [comment]
Yes, that is pretty unhelpful for it act in this way.  But when the client loses connection to the mongodb server, isn't that a bigger problem anyway?  I mean, there isn't really any point in displaying pages not requiring mongodb.  (I realize you're trying to give an example of the problem; I'm trying to determine what needs to be fixed here, and how critical it is.)  

How often is this happening in the cloud?  

Shouldn't the client try first to restore the mongodb connection, and when it fails try a different connection, and then when that fails display a more helpful error/kill itself?

**Comment by edgarcosta** [comment]
Further, this needs to be fixed at some point, but not by May 10.

It is also related to #1165

**Comment by edgarcosta** [comment]
Yes, it's way bigger problem to lose the connection to mongodb. However informing the user that something went wrong is much better than just displaying "Internal Server Error" in big letters.

Not critical at all. 

It hasn't yet happened in the cloud, except when the client by itself was closing the connection, see #1201 

Yes, if we loose the connection we will keep retrying forever and ever. Basically the client dies immediately, but we launch another one right away.

**Comment by edgarcosta** [comment]
More on this, from Sarunas, I bolded the key part:

> Organizational matters aside (not less important, I just don't feel that
> I have been with the project long enough), I'd like to chime in on
> today's interruption. It was about MongoDB underperforming while under
> index creation task, and the Sage clients not handling a slow-responding
> DB backend gracefully. MongoDB replica members, m0 and m1, didn't crash,
> but apparently became “disk-bound” (judging by Munin monitoring data
> https://math.dartmouth.edu/munin/LMFDB/comparison-day.html). I
> understand you have been running a respectably-sized MongoDB for quite a
> while — so is this a known condition? If so, shouldn't index creation be
> done more cautiously, perhaps with some advance warning/consensus,
> perhaps with role-based user authorization? **To complicate matters,
> architecture of Sage client doesn't seem to allow for a graceful
> handling of DB connection timeouts: in my opinion, web page framework
> should be loaded first, independently of even the existence of DB. Then,
> if there are any issues with whatever resources/connections are
> available or not, that can be reported in the main content `<div> `as a
> more-or-less meaningful notice. This would not only help with
> non-responding DB, but with slower tasks as well. Obviously, I have no
> idea how much of a change to the current client would that mean.**
> 
> Sage clients' beta and “production” separation might be trivial using
> Git and separate hosts or containers. I don't know how can a writeable
> beta-MongoDB and a stable (read-only?) DB can be organized. Can Mongo
> replication be suspended for an arbitrary time? How does one ensure that
> changes to beta-DB do not overflow oplog capacity?
> 
> Google Cloud hosting was somewhat of a learning curve to Edgar and me,
> and there still is a lot to learn. When all 8 clients stopped responding
> just minutes before John Voight's talk, Edgar came up with a cloud-based
> solution momentarily. Audience here never even noticed there was an
> issue. So kudos to Edgar! I'm a former experimental physicists and still
> prefer assembling my own server hardware, but solution this quick was
> probably only possible with seemingly unlimited virtual resources of the
> “cloud.”

**Comment by JohnCremona** [comment]
Thanks, that was a very useful description of the situation and answers some questions the west coast team were having already, though not all (since it raises some new questions).
I was going to suggest almost a reverse of the current set-up as follows.  Currently we have the replicated triple {w,m0,m1} with w=warwick as master, and also ms which is not in the replica set but updated somehow once a day.  The cloud webserver gets its dats via the load balancer from one of m0, m1, ms with m0 & m1 having high priority and ms lowest priority.  [Corrections to this description are welcome!]
Instead:  let the cloud webserver use the "static" ms as highest priority so that it will never be affected by the tunnel to warwick or other such issues as happened yesterday.  The replication triple {w,m0,m1} remains (perhaps not both m0 and m1 now needed)  so that development work affecting the master w get replicated to the cloud (including knowl edits).  Once a day the static ms gets updated from m0.
Positive features: cloud webserver not affected by development work on the db, and not subject to any tunnelling delays.
Negative feature: cloud webserver's data up to a day out of date (including knowl content).
Developers editing knowls would do that on beta.lmfdb.org which as now connects to database w.

Would that work, or something like it?

**Comment by davidfarmer** [comment]
I think it is okay to have the data on the public version be one day old.
It is worth that cost to be able to develop in the usual way.

Today we had the unusual situation of needing a particular knowl to
be up-to-date.  That is unlikely to happen routinely.

**Comment by AndrewVSutherland** [comment]
I agree that (almost always) being a day behind is fine.  My concern is having lmfdb.org simultaneously serving data from databases that are out of sync. With the proposed configuration you could potentially (depending on the load) have information that, from the user's perspective, was disappearing and reappearing randomly (and non-reproducibly), which would be very confusing.  I think it would be better to have one of the cloud machines synced with Warwick and used only as a mechanism for transferring data once a day, not as a data source for serving live pages.

I'll note that we have strayed from the original topic of this issue, but I think the issue we are discussing (developers not modifying data that is being used to serve live data) is more important and needs to be dealt with regardless.

**Comment by JohnCremona** [comment]
@AndrewVSutherland You didn't quite understand what I said (or meant to say) -- in my suggestion the only database instances which www.lmfdb.org itself reads from are the statics ones which only change once a day.  Anyone going to beta.lmfdb.org and seeing something different should be surprised.  Does that make sense?

**Comment by AndrewVSutherland** [comment]
@JohnCremona I understand that www.lmfdb.org will be serving from static data on ms, while beta.lmfdb.org will be serving dynamic data from the replication set {w,m0,m1}, and these may differ, and this is fine, no one should expect www.lmfdb.org and beta.lmfdb.org to look the same.  But if I understood what you wrote correctly, you want the load balancer to be spreading work over all 3 of m0,m1,ms, with ms as priority.  This would mean a brief burst of traffic could send a user on lmfdb.org from ms to m0 or m1.  Or do I misunderstand?

**Comment by JohnCremona** [comment]
Perhaps I am not using the correct terminology when I said "load balancing".  I mean that www.lmfdb.org should 100% of its data from one (or more) statics db servers.
That may mean that there is no point in having two replicas in  the cloud (m0 and m1) since their only function would be to incrementally update themselves with whatever changes are being made on the master.  Instead you may have more than one static one -- but that is an implementation detail.

**Comment by edgarcosta** [comment]
Please let's create a a separate issue where we discuss the workflow of the data.
and let's stay on topic.

**Comment by AndrewVSutherland** [comment]
@edgarcosta what exactly needs to be done here?  Would it be enough to just tweak the auto reconnect settings in https://github.com/LMFDB/lmfdb/blob/master/lmfdb/base.py? (see lines 72-103).  It currently it retries 10 times with delay of 1s, which seems way too few (and I don't see why we couldn't just have it retry forever, kill it if you want it dead).

Keep in mind that when you and Sarunas use terms like "client development" most of the lmfdb devs have no idea what you are talking about (from their perspective the lmfdb is a server not client :)), and I'm sure you are not saying that you want the lmfdb application code to actually handle mogno db errors wherever they might occur (there is no need to do that, an individual page request can and should be allowed to fail, you don't want to start wrapping try/excepts around everything).

**Comment by edgarcosta** [comment]
The code never runs that retry code.

It tries only once, and it it fails, raises an exception, for example a pymongo.errors.ServerSelectionTimeoutError.
This exception should be handled by the ```@app.errorhandler(500)```, however that also depends on the db connection...

Further, the current default timeout time is 30s, and by that time the worker already got killed by gunicorn (we should perhaps change the default to the 1-5s range)

Moreover, I tried today to take care of this a bit, but I found it very hard, and after 1h I was touching more code than I felt comfortable with,.

> you don't want to start wrapping try/excepts around everything

Everything is already wrapped in a big try/except, but given that our handler also depends on the db, when the DB fails, everything fails with an "Internal Server Error"

Perhaps, I should close down the issue and create a new one with the title:
"Remove DB dependence of @app.errorhandler(500)"

There are two solutions for this issue, we rewrite the 500.html template to become a very simple one, or we make certain things not depend on the db.

I would prefer the later, for this we would need at least to take care of the user and knowledge package to be able to survive without the DB. (that amounts more or less to make every query to return empty)
The user part I think I could take care of it, however I don't have a clue how to deal with the knowledge package.

To finish, there are clearly two things that I think that should not depend on the DB:
1- 
```
<meta name="description" content="Welcome to the LMFDB, the database of L-functions, modular forms, and related objects. These pages are intended to be a modern handbook including tables, formulas, links, and references for L-functions and their underlying objects.">
```
2- the sidebar

**Comment by edgarcosta** [comment]
Should we change the timeout for pymongo from the default 30s, to something on the range of 1s?
We could set this on `mongoclient.config`

**Comment by AndrewVSutherland2** [comment]
I think 1s might be too low on warwick (especially when backups are
happening or btrfs is cleaning up).  But we could try lowering it in stages
(e.g. try 10s, then 5s) and see how it goes.

How would this help to address this issue? (I'm not disagreeing, just
trying to understand)

**Comment by AndrewVSutherland** [comment]
@edgarcosta I am working on this.

**Comment by AndrewVSutherland** [comment]
#1730 is now live on prod, as soon as we get a chance to test this (which I propose to do by disonnecting one of the two LMFDB servers from the mongo db) I will close this issue.

**Comment by edgarcosta** [comment]
I will take care of this

**Comment by edgarcosta** [comment]
It works, we just need to change a 500 to 503, and reduce the timeout.
I will take care of both.

![screen shot 2016-07-14 at 12 30 11](https://cloud.githubusercontent.com/assets/322822/16826241/12441666-49bf-11e6-9cdc-ecb373aa1059.png)

**Comment by edgarcosta** [comment]
Ups, Drew already took care of the 503.
I will need to wait for #1730 get to prod, to reduce the timeout time.

**Comment by edgarcosta** [comment]
Yes. I have just tested it with a `serverSelectionTimeoutMS = 10000` and everything worked as expected.

Now, lmfdb.org is running with:
```
# Controls how long (in milliseconds) the driver will wait during server monitoring when connecting a new socket to a server before concluding the server is unavailable. Defaults to 20000 (20 seconds).
connectTimeoutMS = 7000

# Controls how long (in milliseconds) the driver will wait to find an available, appropriate server to carry out a database operation; while it is waiting, multiple server monitoring operations may be carried out, each controlled by connectTimeoutMS. Defaults to 30000 (30 seconds).
serverSelectionTimeoutMS = 10000
```
